pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/dictionary/MutableDictionaryReader.java

```diff
@@ -186,4 +202,63 @@ public void print() {
   public boolean isEmpty() {
     return dictionaryIdBiMap.isEmpty();
   }
+
+  /**
+   * Checks if the given value lies within one of the partition ranges. If the value is not within
+   * any of the partition ranges, partitioning is dropped. The effect of that is:
+   * <ul>
+   *   <li> Subsequent values will not be checked against partition ranges. </li>
+   *   <li> Partition information will not be written out to the metadata for this column. </li>
+   * </ul>
+   *
+   * @param value Column value to check.
+   */
+  protected void checkPartition(Object value) {
+    if (partitionFunction != null) {
+       int partition = partitionFunction.getPartition(value);
+
+      for (IntRange partitionValue : partitionValues) {
+        if (partition >= partitionValue.getMinimumInteger() && partition <= partitionValue.getMaximumInteger()) {
+          return;
+        }
+      }
+
+      LOGGER.warn("Column '{}' not partitioned as specified for value: '{}', dropping partition metadata.", column,
```

Comment:
Try to see if you can get the segment name in the log message. Otherwise, with multiple segments being consumed simultaneously, you will have no way of knowing which segment did not fall into the partition correctly.
Alternatively, add a log when the segment is being built as to whether the partition column fitted right or not.